### PR TITLE
Install command

### PR DESCRIPTION
**Pull request by VladimirZaets**
## Description
This pull request adds new commands to allow the developer to work with the api-mesh sources catalog through CLI.

API mesh Source is an entity that declares GraphQL schema and maps it to 3rd party APIs or Runtime Actions.API Mesh config that can be meshed into the user's Mesh schema.

The source is an entity that declares GraphQL schema and maps it to 3rd party APIs or Runtime Actions. GraphQL Mesh config that can be meshed into the user's Mesh schema. To learn more about how GraphQL API Gateway manages configs.

This PR contain the command to install the source to customer API mesh. (aio api-mesh:source:install). Also this PR provide possibility to pass variables in the source. 

## Motivation and Context
To learn more about api-mesh sources functionality, please check this wiki document. https://wiki.corp.adobe.com/display/dxdevx/API+mesh+connector%28sources%29+Customer+Experience

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/commands/api-mesh/source/install.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+					const splited = val.split('=');
+					obj[splited[0].trim()] = splited[1].trim();
+					return obj;
+			  }, {})
+			: {};
+		if (filepath) {
+			try {
+				variables = { ...variables, ...JSON.parse(await readFile(filepath, 'utf8')) };
+			} catch (e) {
+				this.error(`Something went wrong when reading variables file.` + `\n${e.message}`);
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:52]:
```suggestion
				this.error(`Something went wrong trying to read the variables file.` + `\n${e.message}`);
```

File: README.md [hunk 2]
```diff
@@ -90,8 +90,12 @@ To submit a new source, please follow the instructions provided in the [Source R
 ## Commands
 
 ```
-aio api-mesh:source:get 
-NAME_OF_THE_SOURCE
+
+aio api-mesh:source:install NAME_OF_THE_SOURCE
```
Comment (roribio) [inline README.md:94]:
```suggestion
aio api-mesh:source:install SOURCE_NAME
```

Comment (roribio) [reply README.md:94]: I would change this for the other instances as well.

File: README.md [hunk 3]
```diff
@@ -90,8 +90,12 @@ To submit a new source, please follow the instructions provided in the [Source R
...
+aio api-mesh:source:install NAME_OF_THE_SOURCE -v VARIABLE_NAME=VARIABLE_VALUE
+aio api-mesh:source:install NAME_OF_THE_SOURCE -f PATH_TO_FILE_WITH_VARIABLES
```
Comment (roribio) [inline README.md:96]:
```suggestion
aio api-mesh:source:install SOURCE_NAME -f VARIABLES_FILE_PATH
```

File: README.md [hunk 4]
```diff
@@ -90,8 +90,12 @@ To submit a new source, please follow the instructions provided in the [Source R
...
+
+aio api-mesh:source:get NAME_OF_THE_SOURCE
 aio api-mesh:source:get NAME_OF_THE_SOURCE@VERSION_OF_THE_SOURCE
```
Comment (roribio) [inline README.md:99]:
```suggestion
aio api-mesh:source:get SOURCE_NAME@SOURCE_VERSION
```

File: src/commands/api-mesh/source/install.js [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+			}
+		}
+
+		let meshId = null;
+		if (!args.source && !flags.source) {
+			this.error(
+				`The "aio api-mesh:source:install" command requires additional parameters` +
+					`\nUse "aio api-mesh:source:install --help" to see parameters information.`,
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:60]:
```suggestion
					`\nUse "aio api-mesh:source:install --help" to see parameter details.`,
```

File: src/commands/api-mesh/source/install.js [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+		}
+		const sources = flags.source ? flags.source : [args.source];
+		const sourceConfigs = [];
+		for (const source of sources) {
+			let [name, version] = source.split('@');
+			const normalizedName = this.normalizeName(name);
+			if (!list[normalizedName]) {
+				this.error(
+					chalk.red(
+						`The source with the name "${name}" doesn't exist.` +
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:77]:
```suggestion
						`The source named "${name}" doesn't exist.` +
```

File: src/commands/api-mesh/source/install.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+				this.error(chalk.red(`${error.message}\n${error.list.map(err => err.message).join('\n')}`));
+			}
+			sourceConfigs.push(JSON.parse(data));
+		}
+
+		try {
+			meshId = await getMeshId(imsOrgId, projectId, workspaceId);
+		} catch (err) {
+			this.error(
+				`Unable to get mesh ID. Please check the details and try again. RequestId: ${global.requestId}`,
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:116]:
```suggestion
				`Unable to get the mesh ID. Please check the details and try again. RequestId: ${global.requestId}`,
```

File: src/commands/api-mesh/source/install.js [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+			);
+		}
+
+		if (!meshId) {
+			this.error(
+				`Unable to get mesh config. No mesh found for Org(${imsOrgId}) -> Project(${projectId}) -> Workspace(${workspaceId}). Please check the details and try again.`,
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:122]:
```suggestion
				`Unable to get the mesh config. No configured mesh found for Org(${imsOrgId}) -> Project(${projectId}) -> Workspace(${workspaceId}). Please check the details and try again.`,
```

File: src/commands/api-mesh/source/install.js [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+			);
+		}
+
+		try {
+			const mesh = await getMesh(imsOrgId, projectId, workspaceId, meshId);
+
+			if (!mesh) {
+				this.error(
+					`Unable to get mesh with the ID ${meshId}. Please check the mesh ID and try again. RequestId: ${global.requestId}`,
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:131]:
```suggestion
					`Unable to get the mesh configuration with ID ${meshId}. Please check the mesh ID and try again. RequestId: ${global.requestId}`,
```

File: src/commands/api-mesh/source/install.js [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+				);
+			}
+			const verifiedSources = this.verifySourceAlreadyExists(
+				mesh.meshConfig.sources,
+				sourceConfigs,
+			);
+			let override = false;
+			if (verifiedSources.installed.length) {
+				override = await promptConfirm(
+					`The next sources are already installed: ${verifiedSources.installed
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:142]:
```suggestion
					`The following sources are already installed: ${verifiedSources.installed
```

File: src/commands/api-mesh/source/install.js [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+			mesh.meshConfig.sources = override
+				? [...mesh.meshConfig.sources, ...verifiedSources.installed, ...verifiedSources.unique]
+				: [...mesh.meshConfig.sources, ...verifiedSources.unique];
+			try {
+				const response = await updateMesh(imsOrgId, projectId, workspaceId, meshId, {
+					meshConfig: mesh.meshConfig,
+				});
+
+				this.log('Successfully updated the mesh with the id: %s', meshId);
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:157]:
```suggestion
				this.log('Successfully updated the mesh configuration ID: %s', meshId);
```

File: src/commands/api-mesh/source/install.js [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+				this.error(
+					`Unable to update the mesh. Please check the mesh configuration file and try again. If the error persists please contact support. RequestId: ${global.requestId}`,
+				);
+			}
+
+			return mesh;
+		} catch (error) {
+			this.log(error.message);
+			this.error(
+				`Unable to get mesh. Please check the details and try again. If the error persists please contact support. RequestId: ${global.requestId}`,
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:172]:
```suggestion
				`Unable to get the mesh configuration. Please check the details and try again. If the error persists please contact support. RequestId: ${global.requestId}`,
```

File: src/commands/api-mesh/source/install.js [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+InstallCommand.flags = {
+	'variable': Flags.string({
+		char: 'v',
+		description: 'Variables required for the source',
+		multiple: true,
+	}),
+	'variable-file': Flags.string({
+		char: 'f',
+		description: 'Path to the file with variables',
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:207]:
```suggestion
		description: 'Variables file path',
```

File: src/commands/api-mesh/source/install.js [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+	}),
+	'ignoreCache': ignoreCacheFlag,
+};
+
+InstallCommand.description = 'Command install the source to your API mesh.';
```
Comment (roribio) [inline src/commands/api-mesh/source/install.js:212]:
```suggestion
InstallCommand.description = 'Command to install the source to your API mesh.';
```

**Review by roribio — DISMISSED** [review]
I left some suggestions to make the messaging more succinct. Everything else looks good to me 👍 💯

**Comment by of0x** [comment]
Please add a short description to the PR. This will let users  understand what is changing.

**Comment by VladimirZaets** [comment]
@of0x done

File: src/commands/api-mesh/source/install.js (hunk 8 above)
Comment (VladimirZaets) [reply src/commands/api-mesh/source/install.js:122]: This error message is copy of message from other API mesh commands. I prefer to have consistent message

File: src/commands/api-mesh/source/install.js (hunk 7 above)
Comment (VladimirZaets) [reply src/commands/api-mesh/source/install.js:116]: This error message is copy of message from other API mesh commands. I prefer to have consistent message

File: src/commands/api-mesh/source/install.js (hunk 9 above)
Comment (VladimirZaets) [reply src/commands/api-mesh/source/install.js:131]: same here

File: src/commands/api-mesh/source/install.js (hunk 11 above)
Comment (VladimirZaets) [reply src/commands/api-mesh/source/install.js:157]: same here

File: src/commands/api-mesh/source/install.js (hunk 12 above)
Comment (VladimirZaets) [reply src/commands/api-mesh/source/install.js:172]: same

File: README.md [hunk 15]
```diff
@@ -90,9 +90,13 @@ To submit a new source, please follow the instructions provided in the [Source R
 ## Commands
 
 ```
-aio api-mesh:source:get 
-NAME_OF_THE_SOURCE
-aio api-mesh:source:get NAME_OF_THE_SOURCE@VERSION_OF_THE_SOURCE
+
+aio api-mesh:source:install SOURCE_NAME
+aio api-mesh:source:install SOURCE_NAME -v VARIABLE_NAME=VARIABLE_VALUE
+aio api-mesh:source:install SOURCE_NAME -f PATH_TO_FILE_WITH_VARIABLES
```
Comment (roribio) [inline README.md:96]:
```suggestion
aio api-mesh:source:install SOURCE_NAME -f VARIABLES_FILE_PATH
```

File: README.md [hunk 16]
```diff
@@ -90,9 +90,13 @@ To submit a new source, please follow the instructions provided in the [Source R
...
+
+aio api-mesh:source:get SOURCE_NAME
+aio api-mesh:source:get SOURCE_NAME@VERSION_OF_THE_SOURCE
```
Comment (roribio) [inline README.md:99]:
```suggestion
aio api-mesh:source:get SOURCE_NAME@SOURCE_VERSION
```

File: src/commands/api-mesh/source/__tests__/install.test.js [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+					`\nENOENT: no such file or directory, open 'notexist.json'`,
+			);
+		});
+	});
+	test('Check executing with invalid variable type', async () => {
+		const path = `-f=${__dirname}/../__fixtures__/variables-file-invalid.json`;
+		await InstallCommand.run(['test-03', path]).catch(err => {
+			expect(err.message).toEqual(
+				chalk.red(
+					`The next variables has invalid type.\nVariable: ENDPOINT_URL\nRequested type: string`,
```
Comment (roribio) [inline src/commands/api-mesh/source/__tests__/install.test.js:103]:
```suggestion
					`The following variables have an invalid type.\nVariable: ENDPOINT_URL\nRequested type: string`,
```

File: src/commands/api-mesh/source/__tests__/install.test.js [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+	});
+	test('Check executing without passing variables', async () => {
+		await InstallCommand.run(['test-03']);
+		expect(promptInput).toHaveBeenCalledTimes(1);
+	});
+	test('Check executing without passing variables and input data', async () => {
+		await InstallCommand.run(['test-03']).catch(err => {
+			expect(err.message).toEqual(
+				chalk.red(
+					`The next variables has invalid type.\nVariable: ENDPOINT_URL\nRequested type: string`,
```
Comment (roribio) [inline src/commands/api-mesh/source/__tests__/install.test.js:116]:
```suggestion
					`The following variables have an invalid type.\nVariable: ENDPOINT_URL\nRequested type: string`,
```

Comment (roribio) [reply src/commands/api-mesh/source/__tests__/install.test.js:116]: Maybe consider assigning this message to a variable

**Review by roribio — APPROVED** [review]
Approved with suggestions 👍 LGTM

File: src/commands/api-mesh/source/install.js [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+						.join(', ')}.
+                    Do you want to override?`,
+				);
+			}
+
+			mesh.meshConfig.sources = override
+				? [...mesh.meshConfig.sources, ...verifiedSources.installed, ...verifiedSources.unique]
+				: [...mesh.meshConfig.sources, ...verifiedSources.unique];
+			try {
+				const response = await updateMesh(imsOrgId, projectId, workspaceId, meshId, {
```
Comment (revanth0212) [inline src/commands/api-mesh/source/install.js:153]: You also need to consider the fact that your sources can also have files and they need to be added to the `meshConfig.files`. Probably right now there must be none but its good to cater for it.

**Review by revanth0212 — CHANGES_REQUESTED** [review]
Nice work @VladimirZaets 

Just some quick pointers. Mostly it looks good and is almost ready to be merged.

File: src/commands/api-mesh/source/install.js (hunk 8 above)
Comment (jhadobe) [reply src/commands/api-mesh/source/install.js:122]: @revanth0212 I believe @VladimirZaets is concerned that the error messages on lines 122, 131, 157 and 172 all use inconsistent terminology when referring to the mesh. (e.g. what is the mesh, vs the mesh configuration, vs the mesh configuration file, etc) Additionally, some of the error messages appear to be designed to communicate the same information, but are worded differently.

File: src/commands/api-mesh/source/install.js (hunk 19 above)
Comment (VladimirZaets) [reply src/commands/api-mesh/source/install.js:193]: done

**Comment by VladimirZaets** [comment]
Hi @revanth0212 . I fixed your requests and also added file merging functionality. Please take a look, thanks

**Review by revanth0212 — APPROVED** [review]
LGTM

Nice work @VladimirZaets 

Please verify the functionality and merge when you are ready. We can create a release once that is done.

**Review by brasewel — APPROVED** [review]
Looks good to me

**Review by nizarn — APPROVED** [review]

**Pull request merged**